### PR TITLE
[Story devtools] Create logs tab

### PR DESCRIPTION
**Pull request by mszylkowski**
Created the tab "Logs" that displays the validation errors

image

Demo at https://stories-demos-matias.web.app/examples/amp-story/interactives.html#development=1&tab=logs

### CODE REVIEW

File: examples/amp-story/interactives.html
Comment (gmajoulet) [inline examples/amp-story/interactives.html:20]: Re: comment on the other PR

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css:20]: Let's keep the scrollbar, people use it

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css:23]: Re: colors, please find something so we can re-use them and don't end up with hundreds of variables

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css:34]: This doesn't look right, can you try to use a class and other CSS rules that would work without the `*` selector? You should be able to use line-height or flexboxes.

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css:50]: How do you know the font for this weight is loaded?

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css:79]: Please use a class selector

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js:40]: Please use `Services.xhrFor()` and its methods instead

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js:106]: Why not private?

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js:150]: Only the `appendChild` statements should be in the mutate

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css
Comment (mszylkowski) [reply extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css:50]: We're loading 400 and 700 in `amp-story-dev-tools`

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (mszylkowski) [reply extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js:150]: Did it, but why is it better to use 50 different mutateElement calls vs using 1 larger call? Sounds counterproductive to me

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css
Comment (mszylkowski) [reply extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css:34]: Used flexbox. Thought vertical-align was more used

**Comment by mszylkowski** [comment]
@gmajoulet fixed all comments. Not the biggest fan of the scrollbar in the middle being so eye-catchy though...

image

**Comment by processprocess** [comment]
Is it possible to host a demo of this?

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (gmajoulet) [reply extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js]: Ideally you'd only have one that does all the `appendChild`

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js:89]: This looks like it should be in the `layoutCallback` as it does quite a bit of work, right? Won't make a difference in your implementation as both will be called at the same time, but just for good practice.

Also, you should consider adding `return` statements, so the `layoutCallback` resolves once the URL is validated.

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js:106]: Please use promise chaining, you can return `response.text()` and have another `then` later on. You save one indentation level

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js:116]: No error catch? :))

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools.js
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools.js:217]: Can you please double check that the build/layoutCallback aren't called since you're hiding them after they're connected to the DOM? I think it should be fine but might be worth inverting both lines for clarity.

Comment (mszylkowski) [reply extensions/amp-story-dev-tools/0.1/amp-story-dev-tools.js:217]: They are called anyways, no matter if you toggle them before appending or after. Assuming the other PR gets merged, we can move this convo to there

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (mszylkowski) [reply extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js]: Done!

**Comment by mszylkowski** [comment]
@processprocess this is the demo https://stories-demos-matias.web.app/examples/amp-story/interactives.html#development=1&tab=logs

Adding it to the PR description as well

**Review by rsimha — DISMISSED** [review]

**Comment by mszylkowski** [comment]
@rsimha We want to be able to run and show results from the amphtml-validator in our DevTools extension (similar to how it's ran when doing `#development=1` on amp pages) but we'd need to set the package `amphtml-validator` as a dependency.

Is there a better way of doing this or are we god?

**Comment by rsimha** [comment]
> We want to be able to run and show results from the amphtml-validator in our DevTools extension (similar to how it's ran when doing `#development=1` on amp pages) but we'd need to set the package `amphtml-validator` as a dependency to import and run it.
> 
> Is there a better way of doing this or are we good?

I just realized this is being added to `dependencies` in `package.json` and not to `devDependencies`. Normally, this is done when the core runtime or an extension needs to ship code that's contained in an external npm package. We typically end up having to list the file(s) being imported to `build-system/compile/sources.js` so that the code is included in the minified binary that we ship.

Might be worth getting a review from someone on the Bento / Performance / Caching working groups to see if there's a better way.

**Comment by mszylkowski** [comment]
Thanks! We want to import the `amphtml-validator` on our component so we can lazy-load it and run it when we trigger the story dev-tools. The code works fine right now besides Travis tests, but maybe there's a better way of doing this. Will ask other people for feedback!

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (gmajoulet) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js:95]: Up to you but I think it makes sense to also return this to resolve the `layoutCallback()` only once the validation is done

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools.js
Comment (gmajoulet) [reply extensions/amp-story-dev-tools/0.1/amp-story-dev-tools.js:217]: `layoutCallback()` should not be called if the element is hidden, this is enforced by AMP. We should find out why it doesn't work and fix it, so we don't validate the story in the background when we're on the preview tab.

**Review by gmajoulet — APPROVED** [review]
LGTM but please get a proper approval for the package.json

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (mszylkowski) [reply extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js:95]: Yes!

**Comment by mszylkowski** [comment]
@jridgewell for `sources.js`

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css
Comment (processprocess) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css:19]: nit:
I wonder if we want this minimum space between the header and content.
Maybe it's a part of `i-amphtml-story-dev-tools-header`.

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js
Comment (processprocess) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.js:145]: a11y: These should be text elements. 
Maybe they could be `spans` with `display: block`.

**Comment by processprocess** [comment]
The bottom margin on `amp-story-dev-tools` is clipping this. 
Maybe this bottom margin should be on a child instead of the outermost wrapper.
Screen Shot 2020-12-04 at 1 31 06 PM

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css
Comment (mszylkowski) [reply extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-logs.css:19]: Now it's in the header padding and the element padding that it needs in this case 👍

**Comment by mszylkowski** [comment]
@processprocess changed the scrollable area to have padding instead of margin so that the scrollbar is against the window border instead of floating in the middle.

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.css
Comment (processprocess) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.css:24]: ?

Comment (mszylkowski) [reply extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.css]: Oops 😂

File: extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.css
Comment (processprocess) [inline extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.css:36]: `30px` could be a variable so it's easier to update / understand.
I see it used here, in chips-container and `tools-tab-preview`.

**Review by processprocess — APPROVED** [review]
Approving with a couple comments.

**Comment by mszylkowski** [comment]
@processprocess Done with the comments, plan to merge when tests pass (currently failing on flaky tests)

**Pull request merged**
